### PR TITLE
VideoSW: Use VideoCommon framework and drop redundant code

### PR DESCRIPTION
**Pull request by degasus**
This PR tries to reuse all rendering code of videosw, but dropping all of the forked fifo processing code. To match up the VideoCommon interface, lots of dummy functions are needed, but they are better than forked code in my opinion.

### CODE REVIEW

File: Source/Core/DolphinWX/Main.cpp [hunk 1]
```diff
@@ -549,14 +548,6 @@ void Host_ConnectWiimote(int wm_idx, bool connect)
...
-		VideoConfigDiag diag((wxWindow*)parent, backend_name, config_name);
-		diag.ShowModal();
-	}
-	else if (backend_name == "Software Renderer")
-	{
-		SoftwareVideoConfigDialog diag((wxWindow*)parent, backend_name, config_name);
-		diag.ShowModal();
-	}
+	VideoConfigDiag diag((wxWindow*)parent, backend_name, config_name);
+	diag.ShowModal();
```
Comment (degasus) [inline Source/Core/DolphinWX/Main.cpp]: RFC: There are lots of options available without any effect. Is it fine as it is, or shall they be hidden in any way?

File: Source/Core/VideoBackends/Software/DebugUtil.cpp [hunk 2]
```diff
@@ -202,7 +193,7 @@ void OnObjectBegin()
 {
 	if (!g_bSkipCurrentFrame)
 	{
-		if (g_SWVideoConfig.bDumpTextures && swstats.thisFrame.numDrawnObjects >= g_SWVideoConfig.drawStart && swstats.thisFrame.numDrawnObjects < g_SWVideoConfig.drawEnd)
+		if (g_ActiveConfig.bDumpTextures /*&& stats.thisFrame.numDrawnObjects >= g_SWVideoConfig.drawStart && stats.thisFrame.numDrawnObjects < g_SWVideoConfig.drawEnd*/)
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/DebugUtil.cpp]: TODO: fix me

File: Source/Core/VideoBackends/Software/DebugUtil.cpp [hunk 3]
```diff
@@ -211,10 +202,10 @@ void OnObjectEnd()
 {
 	if (!g_bSkipCurrentFrame)
 	{
-		if (g_SWVideoConfig.bDumpObjects && swstats.thisFrame.numDrawnObjects >= g_SWVideoConfig.drawStart && swstats.thisFrame.numDrawnObjects < g_SWVideoConfig.drawEnd)
-			DumpEfb(StringFromFormat("%sobject%i.png",
-						File::GetUserPath(D_DUMPFRAMES_IDX).c_str(),
-						swstats.thisFrame.numDrawnObjects));
+//		if (g_ActiveConfig.bDumpObjects && stats.thisFrame.numDrawnObjects >= g_SWVideoConfig.drawStart && stats.thisFrame.numDrawnObjects < g_SWVideoConfig.drawEnd)
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/DebugUtil.cpp]: TODO: fix me. Are ini options fine? They were exposed within the GUI before.

File: Source/Core/VideoBackends/Software/SWRenderer.cpp [hunk 4]
```diff
@@ -9,152 +9,47 @@
...
 #include "VideoCommon/OnScreenDisplay.h"
-
-static GLuint s_RenderTarget = 0;
-
-static GLint attr_pos = -1, attr_tex = -1;
-static GLint uni_tex = -1;
-static GLuint program;
+#include "VideoCommon/VideoConfig.h"
 
 static u8 *s_xfbColorTexture[2];
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWRenderer.cpp]: Only one is used at all, and this should be moved to virtual XFB code. But this is in my opinion not needed for this PR.

File: Source/Core/VideoBackends/Software/SWRenderer.h [hunk 5]
```diff
@@ -6,24 +6,56 @@
...
+	static u8* GetNextColorTexture();
+	static u8* GetCurrentColorTexture();
 	void SwapColorTexture();
 	void UpdateColorTexture(EfbInterface::yuv422_packed *xfb, u32 fbWidth, u32 fbHeight);
-	void DrawTexture(u8 *texture, int width, int height);
 
-	void Swap(u32 fbWidth, u32 fbHeight);
-	void SwapBuffer();
-}
+	void SetColorMask() override {};
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWRenderer.h]: Lots of dummy functions, but I'd like to clean them up in a later PR.

File: Source/Core/VideoBackends/Software/Tev.cpp [hunk 6]
```diff
@@ -8,18 +8,18 @@
...
+
 #include "VideoCommon/BoundingBox.h"
+#include "VideoCommon/Statistics.h"
+#include "VideoCommon/XFMemory.h"
 
-#ifdef _DEBUG
-#define ALLOW_TEV_DUMPS 1
-#else
+//#ifdef _DEBUG
+//#define ALLOW_TEV_DUMPS 1
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/Tev.cpp]: TODO: Fixme. Also here, INI or GUI option?

File: Source/Core/DolphinWX/Main.cpp (hunk 1 above)
Comment (phire) [inline Source/Core/DolphinWX/Main.cpp]: I think they should be disabled (grayed out)

File: Source/Core/VideoBackends/Software/DebugUtil.cpp (hunk 3 above)
Comment (Tilka) [inline Source/Core/VideoBackends/Software/DebugUtil.cpp]: Yes.

File: Source/Core/VideoBackends/Software/SWRenderer.h [hunk 7]
```diff
@@ -6,22 +6,56 @@
...
+
+	TargetRectangle ConvertEFBRectangle(const EFBRectangle& rc) override;
+
+	void SwapImpl(u32 xfbAddr, u32 fbWidth, u32 fbStride, u32 fbHeight, const EFBRectangle& rc, float Gamma) override;
+
+	void ClearScreen(const EFBRectangle& rc, bool colorEnable, bool alphaEnable, bool zEnable, u32 color, u32 z) override;
+
+	void ReinterpretPixelData(unsigned int convtype) override {}
+
+	bool SaveScreenshot(const std::string &filename, const TargetRectangle &rc) override { return true; };
```
Comment (Tilka) [inline Source/Core/VideoBackends/Software/SWRenderer.h]: Ampersand to the type.

File: Source/Core/VideoBackends/Software/SWVertexLoader.cpp [hunk 8]
```diff
@@ -7,22 +7,41 @@
...
 #include "VideoCommon/VertexLoaderBase.h"
 #include "VideoCommon/VertexLoaderManager.h"
 #include "VideoCommon/VertexLoaderUtils.h"
+#include "VideoCommon/XFMemory.h"
 
+class NullNativeVertexFormat : public NativeVertexFormat
+{
+public:
+	NullNativeVertexFormat() {}
+	void Initialize(const PortableVertexDeclaration &_vtx_decl) override { vtx_decl = _vtx_decl;}
```
Comment (Tilka) [inline Source/Core/VideoBackends/Software/SWVertexLoader.cpp]: Same here.

File: Source/Core/VideoBackends/Software/SWmain.cpp [hunk 9]
```diff
@@ -51,6 +56,86 @@ static volatile struct
...
+			bool isIntensity, bool scaleByHalf, unsigned int cbufid,
+			const float *colmat) override
+		{
+			EfbCopy::CopyEfb();
+		}
+
+		void CopyRectangleFromTexture(
+			const TCacheEntryBase* source,
+			const MathUtil::Rectangle<int> &srcrect,
+			const MathUtil::Rectangle<int> &dstrect) override {}
```
Comment (Tilka) [inline Source/Core/VideoBackends/Software/SWmain.cpp]: And here.

File: Source/Core/VideoBackends/Software/SWRenderer.cpp [hunk 10]
```diff
@@ -138,25 +106,105 @@ void SWRenderer::UpdateColorTexture(EfbInterface::yuv422_packed *xfb, u32 fbWidt
...
 	{
-		std::lock_guard<std::mutex> lk(s_criticalScreenshot);
-		TextureToPng(GetCurrentColorTexture(), fbWidth * 4, s_sScreenshotName, fbWidth, fbHeight, false);
-		// Reset settings
-		s_sScreenshotName.clear();
-		s_bScreenshot.store(false);
+
+		if (g_ActiveConfig.bUseXFB)
+		{
+			EfbInterface::yuv422_packed *xfb = (EfbInterface::yuv422_packed *) Memory::GetPointer(xfbAddr);
```
Comment (Tilka) [inline Source/Core/VideoBackends/Software/SWRenderer.cpp]: Asterisk to the type.

File: Source/Core/VideoBackends/Software/SWmain.cpp [hunk 11]
```diff
@@ -51,6 +56,86 @@ static volatile struct
...
+		TCacheEntry(const TCacheEntryConfig& _config) : TCacheEntryBase(_config) {}
+		~TCacheEntry() {}
+
+		void Load(unsigned int width, unsigned int height,
+			unsigned int expanded_width, unsigned int level) override {}
+
+		void FromRenderTarget(u8* dst, unsigned int dstFormat, u32 dstStride,
+			PEControl::PixelFormat srcFormat, const EFBRectangle& srcRect,
+			bool isIntensity, bool scaleByHalf, unsigned int cbufid,
+			const float *colmat) override
```
Comment (Tilka) [inline Source/Core/VideoBackends/Software/SWmain.cpp]: Meep.

File: Source/Core/VideoBackends/Software/SWmain.cpp [hunk 12]
```diff
@@ -51,6 +56,86 @@ static volatile struct
...
+class XFBSource : public XFBSourceBase
+{
+	void DecodeToTexture(u32 xfbAddr, u32 fbWidth, u32 fbHeight) override {}
+	void CopyEFB(float Gamma) override {}
+};
+
+class FramebufferManager : public FramebufferManagerBase
+{
+	XFBSourceBase* CreateXFBSource(unsigned int target_width, unsigned int target_height, unsigned int layers) override { return new XFBSource(); }
+	void GetTargetSize(unsigned int *width, unsigned int *height) override {};
```
Comment (Tilka) [inline Source/Core/VideoBackends/Software/SWmain.cpp]: Moop.

File: Source/Core/VideoBackends/Software/EfbCopy.cpp [hunk 13]
```diff
@@ -33,31 +31,9 @@ namespace EfbCopy
 		INFO_LOG(VIDEO, "xfbaddr: %x, fbwidth: %i, fbheight: %i, source: (%i, %i, %i, %i), Gamma %f",
 				 xfbAddr, fbWidth, fbHeight, sourceRc.top, sourceRc.left, sourceRc.bottom, sourceRc.right, Gamma);
 
-		if (!g_SWVideoConfig.bBypassXFB)
-		{
-			EfbInterface::yuv422_packed* xfb_in_ram = (EfbInterface::yuv422_packed *) Memory::GetPointer(xfbAddr);
+		EfbInterface::yuv422_packed* xfb_in_ram = (EfbInterface::yuv422_packed *) Memory::GetPointer(xfbAddr);
```
Comment (Tilka) [inline Source/Core/VideoBackends/Software/EfbCopy.cpp]: Blub.

**Comment by waddlesplash** [comment]
Just curious - does this affect performance at all? Or is VideoCommon's FIFO processor about the same?

**Comment by degasus** [comment]
VideoCommon fifo processing is a bit faster, but this is really nothing compared to rasterization + TEV +  ...

**Comment by JMC47** [comment]
This PR breaks Medabots Infinity on SW Renderer, where as it previously worked.

**Comment by JMC47** [comment]
Cel-damage regresses in this PR on Software renderer.  Instead of all the characters being too dark, they now look like they have tire tracks over them.

**Comment by JMC47** [comment]
In this Pull Request, Ed, Edd, Eddy Misedventures lighting works in Software Renderer, but not hardware renderers.

**Comment by JMC47** [comment]
This also reveals that the upside down A/B buttons in Mario Golf DO NOT RENDER AT ALL in the software renderer.

**Comment by JMC47** [comment]
This reveals that the Sonic Adventure GUI does render on Software Renderer, but not on hardware backends in some areas.

**Comment by JMC47** [comment]
This PR also reveals that Fortune Street's white box glitch does not happen on Software renderer (software renderer would previously crash before loading a level, but now I can use savestates.)

**Comment by degasus** [comment]
Medabots Infinity crash is fixed in PR #3268.

File: Source/Core/VideoBackends/Software/SWRenderer.h [hunk 14]
```diff
@@ -6,22 +6,56 @@
...
+	void SetColorMask() override {};
+	void SetBlendMode(bool forceUpdate) override {};
+	void SetScissorRect(const EFBRectangle& rc) override {};
+	void SetGenerationMode() override {};
+	void SetDepthMode() override {};
+	void SetLogicOpMode() override {};
+	void SetDitherMode() override {};
+	void SetSamplerState(int stage, int texindex, bool custom_tex) override {};
+	void SetInterlacingMode() override {};
+	void SetViewport() override {};
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWRenderer.h]: Just write a default implementation of this functions instead.

File: Source/Core/VideoBackends/Software/SWRenderer.h [hunk 15]
```diff
@@ -6,22 +6,56 @@
...
+	void RenderText(const std::string& pstr, int left, int top, u32 color) override;
+	u32 AccessEFB(EFBAccessType type, u32 x, u32 y, u32 poke_data) override;
+
+	u16 BBoxRead(int index) override;
+	void BBoxWrite(int index, u16 value) override;
+
+	int GetMaxTextureSize() override { return 16 * 1024; };
+
+	void ResetAPIState() override {};
+	void RestoreAPIState() override {};
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWRenderer.h]: This should not be within videocommon API at all.

File: Source/Core/VideoBackends/Software/SWVertexLoader.cpp [hunk 16]
```diff
@@ -7,22 +7,41 @@
...
+{
+public:
+	NullNativeVertexFormat() {}
+	void Initialize(const PortableVertexDeclaration& _vtx_decl) override { vtx_decl = _vtx_decl;}
+	void SetupVertexPointers() override {}
+};
+
+NativeVertexFormat* SWVertexLoader::CreateNativeVertexFormat()
+{
+	return new NullNativeVertexFormat;
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWVertexLoader.cpp]: Is it possible to return nullptr?

File: Source/Core/VideoBackends/Software/SWmain.cpp [hunk 17]
```diff
@@ -51,6 +56,90 @@ static volatile struct
...
+
+class XFBSource : public XFBSourceBase
+{
+	void DecodeToTexture(u32 xfbAddr, u32 fbWidth, u32 fbHeight) override {}
+	void CopyEFB(float Gamma) override {}
+};
+
+class FramebufferManager : public FramebufferManagerBase
+{
+	XFBSourceBase* CreateXFBSource(unsigned int target_width, unsigned int target_height, unsigned int layers) override { return new XFBSource(); }
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWmain.cpp]: Is it possible to just return nullptr?

File: Source/Core/VideoBackends/Software/SWmain.cpp [hunk 18]
```diff
@@ -51,6 +56,90 @@ static volatile struct
...
+		void Bind(unsigned int stage) override {}
+
+		bool Save(const std::string& filename, unsigned int level) override { return false; }
+	};
+
+	~TextureCache() {}
+
+	TCacheEntryBase* CreateTexture(const TCacheEntryConfig& config) override
+	{
+		return new TCacheEntry(config);
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWmain.cpp]: Is it possible to return nullptr?

File: Source/Core/VideoBackends/Software/SWmain.cpp [hunk 19]
```diff
@@ -162,186 +218,59 @@ void VideoSoftware::Shutdown()
...
+	g_perf_query = new PerfQuery();
+	Fifo_Init(); // must be done before OpcodeDecoder_Init()
+	OpcodeDecoder_Init();
+	IndexGenerator::Init();
+	VertexShaderManager::Init();
+	PixelShaderManager::Init();
+	g_texture_cache = new TextureCache();
+	SWRenderer::Init();
+	VertexLoaderManager::Init();
+	g_framebuffer_manager = new FramebufferManager();
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/SWmain.cpp]: May this variable be nullptr?

**Comment by JMC47** [comment]
What did you recently change?  sms grey cubes is fixed.

**Comment by degasus** [comment]
I've dropped the removal of the videosw config gui, so there is no need to discuss about the videocommon GUI way in this PR.

**Comment by degasus** [comment]
Half of the toggled fifoci runs are because of broken fifo logs (see PR #3401), the other ones seems to be better. What do you think about merging as-it-is? It looks like I don't get myself motivated enough to spend hours on already fixed bugs :/

**Pull request merged**